### PR TITLE
nixos: compress make-ext4-fs with zstd

### PR DESCRIPTION
**Pull request by lovesegfault**
###### Motivation for this change
The AArch64 build has been failing for a few days with the output of it's ext4 image being >2.0GB. This compresses it with bz2, thus solving the problem.

I'll add that currently we compress all these large images with bz2, which is very slow, I'd like to suggest a change to zstd or lzo. This would be specially meaningful in architectures where usually there are more cores at the expense of clock (since compression/decompression is single threaded.) In particular ARM families would benefit, with no downside to other arches. I am preparing this in another branch, but would like to see this one merged first. 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @samueldr @andir @grahamc @cleverca22

### CODE REVIEW

**Comment by lovesegfault** [comment]
Difference between bzip2 and zstd
```
❯ time bzip2 -k nixos-sd-image-20.03pre204199.3140fa89c51-aarch64-linux.img
bzip2 -k nixos-sd-image-20.03pre204199.3140fa89c51-aarch64-linux.img  145.93s user 0.44s system 99% cpu 2:26.39 total
❯ time zstd nixos-sd-image-20.03pre204199.3140fa89c51-aarch64-linux.img
nixos-sd-image-20.03pre204199.3140fa89c51-aarch64-linux.img : 26.99%   (2189545472 => 591009441 bytes, nixos-sd-image-20.03pre204199.3140fa89c51-aarch64-linux.img.zst)
zstd nixos-sd-image-20.03pre204199.3140fa89c51-aarch64-linux.img  6.50s user 0.33s system 110% cpu 6.196 total
```

**Comment by lovesegfault** [comment]
I've done the work to use zstd in another branch and the difference is staggering.

**Comment by lovesegfault** [comment]
```
[105 built, 33 copied (814.4 MiB), 187.4 MiB DL]
https://github.com/NixOS/nixpkgs/pull/75595
55 package were built:
vimPlugins.UltiSnips vimPlugins.agda-vim vimPlugins.airline vimPlugins.calendar vimPlugins.coc-lists vimPlugins.coc-nvim vimPlugins.coc-pairs vimPlugins.coc-rls vimPlugins.coc-snippets vimPlugins.coc-tabnine vimPlugins.coc-tsserver vimPlugins.coc-yank vimPlugins.csv vimPlugins.denite vimPlugins.deoplete-fish vimPlugins.deoplete-jedi vimPlugins.deoplete-lsp vimPlugins.deoplete-nvim vimPlugins.dhall-vim vimPlugins.echodoc vimPlugins.editorconfig-vim vimPlugins.falcon vimPlugins.fugitive vimPlugins.gentoo-syntax vimPlugins.gv-vim vimPlugins.jedi-vim vimPlugins.lightline-vim vimPlugins.ncm2-ultisnips vimPlugins.neosnippet vimPlugins.neoterm vimPlugins.neoyank-vim vimPlugins.polyglot vimPlugins.quick-scope vimPlugins.quickfix-reflector-vim vimPlugins.seoul256-vim vimPlugins.table-mode vimPlugins.targets-vim vimPlugins.tmuxNavigator vimPlugins.vim-airline-themes vimPlugins.vim-flagship vimPlugins.vim-go vimPlugins.vim-grepper vimPlugins.vim-jsx-pretty vimPlugins.vim-mucomplete vimPlugins.vim-orgmode vimPlugins.vim-peekaboo vimPlugins.vim-pencil vimPlugins.vim-plug vimPlugins.vim-ruby vimPlugins.vim-slime vimPlugins.vim-snippets vimPlugins.vim-tmux-focus-events vimPlugins.vim-trailing-whitespace vimPlugins.vimtex vimPlugins.zig-vim
```

File: nixos/lib/make-ext4-fs.nix [hunk 1]
```diff
@@ -21,17 +22,22 @@ let
...
+pkgs.stdenv.mkDerivation rec {
+  name = "ext4-fs.img.bz2";
 
-  nativeBuildInputs = [e2fsprogs.bin libfaketime perl lkl];
+  nativeBuildInputs = [ bzip2 e2fsprogs.bin libfaketime perl lkl ];
 
   buildCommand =
     ''
+      # foo.img.bz2
+      local name="${name}"
```
Comment (samueldr) [inline nixos/lib/make-ext4-fs.nix]: Setting the `name` attribute already makes `name` available as an environment variable. Those lines can be removed.

I believe this also removes the need for `rec`.

File: nixos/lib/make-ext4-fs.nix [hunk 2]
```diff
@@ -21,17 +22,22 @@ let
...
+      # foo.img
+      local img="''${name%.*}"
```
Comment (samueldr) [inline nixos/lib/make-ext4-fs.nix]: The name of the temp image is not important at all. We could just as well call it `tmp.img` outright during the full build. There is no need to make it generic from the name of the derivation. `$out` is already doing all of the actually important naming for us :).

File: nixos/lib/make-ext4-fs.nix [hunk 3]
```diff
@@ -42,38 +48,39 @@ pkgs.stdenv.mkDerivation {
...
-      cptofs -t ext4 -i $out $storePaths /nix/store/
+      cptofs -t ext4 -i ./$img $storePaths /nix/store/
 
       (
       echo "copying files to image..."
-      cd ./files
-      cptofs -t ext4 -i $out ./* /
+      pushd ./files
+      cptofs -t ext4 -i ../$img ./* /
+      popd
```
Comment (samueldr) [inline nixos/lib/make-ext4-fs.nix]: Revert to `cd ./files`; the subshell is safer than `pushd/popd` here. The subshell (marked by the parentheses) does not change the PWD outside of the parentheses.

This change is basically a no-op adding a line to the code :).

File: nixos/lib/make-ext4-fs.nix [hunk 4]
```diff
@@ -82,10 +89,12 @@ pkgs.stdenv.mkDerivation {
...
         echo "Resizing from $blocks blocks to $size blocks. (~ $((size*blocksize/1024/1024))MiB)"
-        EXT2FS_NO_MTAB_OK=yes resize2fs $out -f $size
+        EXT2FS_NO_MTAB_OK=yes resize2fs ./$img -f $size
       )
 
       # And a final fsck, because of the previous truncating.
-      fsck.ext4 -n -f $out
+      fsck.ext4 -n -f ./$img
+
+      bzip2 -c ./$img > $out
```
Comment (samueldr) [inline nixos/lib/make-ext4-fs.nix]: Could we add a `echo "Compressing image..."` so the end-user knows it's not stuck at fsck? Especially considering the length of time it's actually compressing.

File: nixos/modules/installer/cd-dvd/sd-image.nix [hunk 5]
```diff
@@ -146,11 +146,14 @@ in
           echo "file sd-image $img" >> $out/nix-support/hydra-build-products
         fi
 
+        root_img="./root-fs.img"
```
Comment (samueldr) [inline nixos/modules/installer/cd-dvd/sd-image.nix]: No real need to make this a variable, as it's basically static and self-contained in the `sd-image.nix` build...

But, no real need to change it either. Just saying.

**Review by samueldr — COMMENTED** [review]
As it is, this breaks the API of `make-ext4-fs.nix`.

I do not know what is our guarantee with internal~ish builders like those.

This could be fixed by adding a `compressed ? false` param to the builder. the main difference would be that `bzip2` invocation is not used, but rather a `cp`.

**Review by samueldr — CHANGES_REQUESTED** [review]
(Ugh, forgot to tick *request changes*)

**Comment by samueldr** [comment]
I don't think that the following is an issue with which we should concern ourselves right now, but to be kept in mind for the eventual rewrite.

Writing a temporary file is costly. It could result in failing builds if the `$NIX_BUILD_TOP` directory ends up in a `tmpfs`, and memory is in contention. A future approach would better be served if it could avoid copying the files and rather directly compressing / streaming in place.

For the second part (streaming into the final image), it would need metadata written in $out (so $out would not be directly an image file) that tells the image builder what is the size of the image. For the first part, (compression) I guess it depends on what compresses the image file. Are there ways that it can be done in-place? Though, if it's writing in a `$out` directory, rather than a `$out` file, the temp file could be in `$out` and removed as needed if this is a concern.

A final strategy for the future could be to merge multiple snippets into one big honking derivation that does all the work. The main drawback would be that the derivation is less flexible, and takes more time to build. A well architected solution could possibly do this with an option.

File: nixos/modules/installer/cd-dvd/sd-image.nix [hunk 6]
```diff
@@ -141,16 +143,19 @@ in
 
         echo "${pkgs.stdenv.buildPlatform.system}" > $out/nix-support/system
         if test -n "$compressImage"; then
-          echo "file sd-image $img.bz2" >> $out/nix-support/hydra-build-products
+          echo "file sd-image $img.zst" >> $out/nix-support/hydra-build-products
```
Comment (samueldr) [inline nixos/modules/installer/cd-dvd/sd-image.nix]: Leftover zstd change still present.

File: nixos/lib/make-ext4-fs.nix [hunk 7]
```diff
@@ -82,10 +87,17 @@ pkgs.stdenv.mkDerivation {
...
       # And a final fsck, because of the previous truncating.
-      fsck.ext4 -n -f $out
+      fsck.ext4 -n -f ./$img
+
+      if [ ${builtins.toString compressImage} ]; then
+        echo "Compressing image"
+        zstd -v ./$img -o $out
+      else
+        cp ./$img $out
+      fi
```
Comment (samueldr) [inline nixos/lib/make-ext4-fs.nix]: I'm not sure it warrants a change, but the condition could have been handled nix-side

```
${if compressImage then ''
  echo "Compressing image"
  zstd -v ./$img -o $out
'' else ''
  cp ./$img $out
''}
```

**Review by samueldr — CHANGES_REQUESTED** [review]

**Review by samueldr — APPROVED** [review]

**Pull request merged**